text/0080-datetime-extension.md

```diff
@@ -0,0 +1,490 @@
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  9 <= ((context.currentTime / (60 * 60)) % 24) &&
+  ((context.currentTime / (60 * 60)) % 24) < 18
+};
+```
+
+Note that the localized version of this example, with timezone offset, could be supported using the `+` or `-` operators on `context.currentTime`.
+
+**Prevent employees from accessing work documents on the weekend.**
+
+With Unix Time, this requires `/` and `%` operators to compute the `dayOfWeek`, which isn't currently expressible in Cedar.
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+With Unix Time, this requires `/` and `%` operators to compute whether or not it is a leap year, which isn't currently expressible in Cedar.
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.currentTime > 1580511600 &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+### Alt. B: Pass results of time checks in the context
+
+Another workaround we have suggested is to simply handle date/time logic _outside_ of Cedar, and pass the results of checks in the context. For example, you could pass in fields like  `context.isWorkingHours` or `context.dayOfTheWeek`.
+
+Here are the previous examples rewritten to use additional context.
+
+**Only allow experienced, tenured persons from the Hardware Engineering department to see prototypes.**
+
+```cedar
+permit(
+  principal is User,
+  action == Action::"view",
+  resource in Folder::"device_prototypes"
+)
+when {
+  principal.department == "HardwareEngineering" &&
+  principal.jobLevel >= 10 &&
+  principal.hireDate <= context.minimumHiringDateForAccess
+};
+```
+
+**Only allow user "alice" to view JPEG photos for one week after creation.**
+
+```cedar
+permit(
+  principal == User::"alice",
+  action == PhotoOp::"view",
+  resource is Photo
+) when {
+  resource.fileType == "JPEG" &&
+  resource.creationDate >= context.oldestViewableDate
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm UTC.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.isWorkingHours
+};
+```
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  context.isTheWeekend
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  context.isMyBirthday && context.isLeapDay
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.afterBrexit &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+
+### Alt. C: Combine Unix Time with additional Pre-Computed context.
```

Comment:
This is my preference. In addition to the concerns Emina raised, I'd like to point out that different languages` time libraries aren't consistent. E.g., Java uses the Java Time-Scale (https://stackoverflow.com/questions/30984599/how-does-the-oracle-java-jvm-know-a-leap-second-is-occurring). UTC requires empirical observations of leap seconds (https://blogs.mathworks.com/loren/2021/05/06/what-is-a-leap-second-anyway/). IMO, Cedar shouldn't have an opinion about which of these you should use.

One can argue that a few seconds of authorization difference don't matter, and I agree this is usually true but I can think of cases like migrating existing authz code written in LangOfChoice to Cedar where these discrepancies will come up in shadow mode. Also I'm uncomfortable with an unbounded divergence between Cedar time and UTC / Java Time / whatever.

I think the right solution is to pass the date information via the context so that it matches the full-fledged date+time facilities of the language you're using.

Reply:
Sort of. What I'd like is for people to pass things like `context.day_of_week` (or better, `context.is_weekday`) and `context.month` in the context and add "native" (or maybe I should call it "extended"?) support just for the comparison operators.

Reply:
Yes, the first 10 operators would be analyzable because they are expressible as simple arithmetic operations over longs.  And since every long is a valid timestamp, we don't need to try and enforce any well-formedness constraints on them (which was the issue with the YYYYMMDDHHSS representation). 

I'll have to think more about the `DT.toDate()` operator.  If there is a definition of this operator that is expressed as arithmetic over longs, we can analyze it.   Defining this operator depends on what "midnight" means.  The usual definition is timezone-specific, so we can't support that one.  But if there is some alternative definition based on epoch arithmetic, it should be expressible in SMT.

Reply:
I think the `toDate` op can just be: `DT / (24*60*60)` which will give you the number of days since Jan 1 1970. When pretty-printing this from Rust, we could calculate the calendar day.

Reply:
My thinking was actually `(DT - (DT / 24*60*60))` -- which gives you the "start of the day." With this you can then do something like:

`resource.createdAt > currentTime.toDate() && resource.createdAt < currentTime.toDate().add(duration(1, 'day'))`

Similar with the `toTime()` which would give you the amount of time since the day started... assuming UTC.

Reply:
(Dropping milliseconds for readability)
`DT / (24*60*60)` should give you the number of days elapsed since epoch start (which will be at midnight). I think that's what `toDate` (or `toDays` or some other name) should do.

To get the "start of day" I think you want `epoch_start + (24*60*60)*(DT / (24*60*60))`

I think you maybe meant `(DT - (24*60*60)*(DT / 24*60*60))` which would give you the duration since "start of day." I'd be open to making that (or preferably "start_of_day") a function as well.

Reply:
Whoops! Yeah! What I actually meant was `(DT - (DT % (24*60*60))` which drops the left over seconds. That one character typo made all the difference.. 😮‍💨 🌴.